lib/posix-fdtab/fdtab.c

```diff
@@ -0,0 +1,482 @@
+static inline struct uk_ofile *ofile_new(struct uk_fdtab *tab)
+{
+	struct uk_ofile *of = uk_malloc(tab->alloc, sizeof(*of));
+
+	if (of)
+		uk_ofile_init(of);
+	return of;
+}
+static inline void ofile_del(struct uk_fdtab *tab, struct uk_ofile *of)
+{
+	uk_free(tab->alloc, of);
+}
+
+static inline void ofile_acq(struct uk_ofile *of)
+{
+	uk_refcount_acquire(&of->refcnt);
+}
+static inline void ofile_rel(struct uk_fdtab *tab, struct uk_ofile *of)
+{
+	if (uk_refcount_release(&of->refcnt)) {
+		uk_file_release(of->file);
+		ofile_del(tab, of);
+	}
+}
+
+#if CONFIG_LIBPOSIX_FDTAB_LEGACY_SHIM
+
+static inline void file_acq(void *p, int flags)
+{
+	if (flags & UK_FDTAB_VFSCORE)
+		fhold((struct vfscore_file *)p);
+	else
+		ofile_acq((struct uk_ofile *)p);
+}
+static inline void file_rel(struct uk_fdtab *tab, void *p, int flags)
+{
+	if (flags & UK_FDTAB_VFSCORE)
+		fdrop((struct vfscore_file *)p);
+	else
+		ofile_rel(tab, (struct uk_ofile *)p);
+}
+
+#else /* !CONFIG_LIBPOSIX_FDTAB_LEGACY_SHIM */
+
+#define file_acq(p, f) ofile_acq((struct uk_ofile *)(p))
+#define file_rel(t, p, f) ofile_rel((t), (struct uk_ofile *)(p))
+
+#endif /* !CONFIG_LIBPOSIX_FDTAB_LEGACY_SHIM */
+
+/* Ops */
+
+int uk_fdtab_open(const struct uk_file *f, unsigned int mode)
+{
+	struct uk_fdtab *tab;
+	struct uk_ofile *of;
+	int flags;
+	const void *entry;
+	int fd;
+
+	UK_ASSERT(f);
+
+	tab = _active_tab();
+	of = ofile_new(tab);
+	if (!of)
+		return -ENOMEM;
+	/* Take refs on file & ofile */
+	uk_file_acquire(f);
+	ofile_acq(of);
+	/* Prepare open file */
+	of->file = f;
+	of->pos = 0;
+	of->mode = mode & ~O_CLOEXEC;
+	/* Place the file in fdtab */
+	flags = (mode & O_CLOEXEC) ? UK_FDTAB_CLOEXEC : 0;
+	entry = fdtab_encode(of, flags);
+	fd = uk_fmap_put(&tab->fmap, entry, 0);
+	if (fd >= UK_FDTAB_SIZE)
+		goto err_out;
+	return fd;
+err_out:
+	/* Release open file & file ref */
+	ofile_rel(tab, of);
+	return -ENFILE;
+}
+
+int uk_fdtab_setflags(int fd, int flags)
+{
+	struct uk_fdtab *tab;
+	struct uk_fmap *fmap;
+	void *p;
+	struct fdval v;
+	const void *newp;
+
+	if (flags & ~O_CLOEXEC)
+		return -EINVAL;
+
+	tab = _active_tab();
+	fmap = &tab->fmap;
+
+	p = uk_fmap_critical_take(fmap, fd);
+	if (!p)
+		return -EBADF;
+	v = fdtab_decode(p);
+	v.flags &= ~UK_FDTAB_CLOEXEC;
+	v.flags |= flags ? UK_FDTAB_CLOEXEC : 0;
+
+	newp = fdtab_encode(v.p, v.flags);
+	uk_fmap_critical_put(fmap, fd, newp);
+	return 0;
+}
+
+int uk_fdtab_getflags(int fd)
+{
+	struct uk_fdtab *tab = _active_tab();
+	void *p = uk_fmap_lookup(&tab->fmap, fd);
+	struct fdval v;
+	int ret;
+
+	if (!p)
+		return -EBADF;
+
+	v = fdtab_decode(p);
+	ret = 0;
+	if (v.flags & UK_FDTAB_CLOEXEC)
+		ret |= O_CLOEXEC;
+	return ret;
+}
+
+#if CONFIG_LIBPOSIX_FDTAB_LEGACY_SHIM
+int uk_fdtab_legacy_open(struct vfscore_file *vf)
+{
+	struct uk_fdtab *tab = _active_tab();
+	const void *entry;
+	int fd;
+
+	fhold(vf);
+	entry = fdtab_encode(vf, UK_FDTAB_VFSCORE);
+	fd = _fmap_put(&tab->fmap, entry, 0);
```

Comment:
I think this was forgotten on the last update to `uk_fmap_put`. I'll do this for you while upstreaming 👍🏻